### ISSUE TITLE
jobstart() uses duplicate job ids before the JobActivity autocommand runs

### ISSUE THREAD

**Issue by benekastah** [neovim/neovim#2162]
This only happens when the first command completes very quickly and the two jobs are started in different event loop cycles from each other.

How to reproduce:

Save the following to `jobprob.vim`
```
au BufReadPost * let job1 = jobstart('job1', 'echo', ['1'])
au FileType * let job2 = jobstart('job2', 'echo', ['2'])
au JobActivity * echom 'job1='.job1.', job2='.job2
```

Then open `nvim` and do:

```
:so jobprob.vim
:e jobprob.vim
```

I get the output: `job1=2, job2=2`.

I think it is important that jobs are uniquely identifiable all the way until the `JobActivity` is fired that notifies that the job has exited. In [neomake](https://github.com/benekastah/neomake), I am using the job id as a key to a dict that has other information I need to properly process the results of a job. If that data is overwritten by a new job before the data from the first job is processed, I have no choice but to ignore the results of the job altogether.

**Comment by splinterofchaos** [comment]
At first glance, this would seem impossible. The algorithm to assign an id to a job:
```c
  // Search for a free slot in the table
  for (i = 0; i < MAX_RUNNING_JOBS; i++) {
    if (table[i] == NULL) {
      break;
    }
  }

  job = xmalloc(sizeof(Job));
  // Initialize
  job->id = i + 1;
```
but we don't actually assign `table[i] = job` until the very end of this function, possibly after one of the other functions has ran the event loop again(?) leading to a second `job_start()` finding the same entry vacant.

This patch may fix it:
```patch
diff --git a/src/nvim/os/job.c b/src/nvim/os/job.c
index 9855a19..610b92e 100644
--- a/src/nvim/os/job.c
+++ b/src/nvim/os/job.c
@@ -115,6 +115,9 @@ Job *job_start(JobOptions opts, int *status)
   job->opts = opts;
   job->closed = false;
 
+  // Save the job to the table
+  table[i] = job;
+
   process_init(job);
 
   if (opts.writable) {
@@ -134,6 +137,7 @@ Job *job_start(JobOptions opts, int *status)
 
   // Spawn the job
   if (!process_spawn(job)) {
+    table[i] = NULL;
     if (opts.writable) {
       uv_close((uv_handle_t *)&job->proc_stdin, close_cb);
     }
@@ -167,8 +171,6 @@ Job *job_start(JobOptions opts, int *status)
     rstream_set_stream(job->err, job->proc_stderr);
     rstream_start(job->err);
   }
-  // Save the job to the table
-  table[i] = job;
 
   return job;
 }
```

**Comment by benekastah** [comment]
I'm not sure if this used to be a problem, but the copy of neovim I am using to reproduce this was built today. I'll try applying your patch sometime soon to see if it works.

**Comment by oni-link** [comment]
I think the problem is that job ids are reused, not the event loop. 
The job id is directly connected to the first free field in the job table (`== job_id - 1`). When a job exits, its field in the job table is reused, and so is the corresponding job id. So the next running job can have the same job id as a previously running job.

**Comment by oni-link** [comment]
Sourcing the following script (`j3.vim`)
```viml
let jobz = jobstart('shellz', 'sleep', ['10']) 
let jobx = jobstart('shellx', 'echo', ['xyz']) 
let joby = jobstart('shelly', 'sleep', ['15']) 
au BufReadPost * let job1 = jobstart("job1", 'echo', ['1'])
au FileType * let job2 = jobstart("job2", 'echo', ['3'])
au JobActivity job* echom 'job1='.job1.', job2='.job2.', jobx='.jobx.', joby='.joby.', jobz='.jobz
e j3.vim
```
gives this output for me:
```
job1=3, job2=3, jobx=3, joby=4, jobz=2
```
The jobs `jobz` and `joby` are long running jobs, so they "block" their id for a long time. `jobx` is started in between this two jobs but exits fast. After that `job1` and `job2` are started and reuse the job id of `jobx`, the first free job id.

**Comment by tarruda** [comment]
@oni-link What cant happen is `job2` having the same id as `job1` before `job1` "exit" event is triggered. Here's what I have tested:

```vim
au BufReadPost * let job1 = jobstart('job1', 'echo', ['1']) | call append(line('$'), 'job1='.job1.' started')
au FileType * let job2 = jobstart('job2', 'echo', ['2']) | call append(line('$'), 'job2='.job2.' started')
au JobActivity job1 if v:job_data[1] == 'exit' | call append(line('$'), 'job1='.job1.' exited') | endif
au JobActivity job2 if v:job_data[1] == 'exit' | call append(line('$'), 'job2='.job2.' exited') | endif
```

and then run `nvim -u NONE --cmd 'syntax on' --cmd 'filetype on' -S jobprob.vim +'e jobprob.vim'` results in a buffer with:

```vim
au BufReadPost * let job1 = jobstart('job1', 'echo', ['1']) | call append(line('$'), 'job1='.job1.' started')
au FileType * let job2 = jobstart('job2', 'echo', ['2']) | call append(line('$'), 'job2='.job2.' started')
au JobActivity job1 if v:job_data[1] == 'exit' | call append(line('$'), 'job1='.job1.' exited') | endif
au JobActivity job2 if v:job_data[1] == 'exit' | call append(line('$'), 'job2='.job2.' exited') | endif
job2=2 started
job1=2 started
job2=2 exited
job1=2 exited
```

Which confirms the bug is there. This is what I imagine is happening:

- `FileType` is triggered and job2 is started
- nvim is notified that job2 exited which schedules a deferred event for processing the corresponding `JobActivity`. At this point the job id is free for reuse.
- `BufReadPost` is triggered and job1 is started reusing job2 id before the exit autocommand was processed

Unfortunately we can't get rid of deferred events yet, so I think the simplest solution here is to not reuse job ids.

**Comment by justinmk** [comment]
Fixed by #2980, per https://github.com/neovim/neovim/pull/2771#issuecomment-122163695
